return plaintext data faster on unpadded decryption

jethrogb opened the pull request:
## Description
When doing symmetric decryption, the code buffers some output data because it might need to strip the padding later. If there is no padding, this buffering is unnecessary. This patch changes the behavior of `mbedtls_cipher_update` so that as much decrypted data as possible is returned.

This change is submitted under CLA with Fortanix, Inc.

## Status
READY

## Requires Backporting
NO

## Migrations
NO

## Additional comments
Documentation for `mbedtls_cipher_update` says that “[`output`] should be able to hold at least ilen + block_size.” This patch does not change this requirement.

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

RonEld commented:
@jethrogb Thank you for your contribution!
The CLA is still under processing.
In the meanwhile, please mention the justification for this PR. I am not sure this change affects throughput of the while data ciphering, only by moving the plaintext from `mbedtls_cipher_finish()` to `mbedtls_cipher_update()`

RonEld commented:
Hi @jethrogb Thank you for your clarification. Unfortunately, the CLA has not been approved yet, due to some legal comments, and Fortanix will need to resign the CLA.

RonEld commented:
BTW, when you mention AF_ALG cipher api, does it mean you intend to use Mbed TLS in kernel space?

jethrogb commented:
If you were to implement the AF_ALG API in your own kernel using mbedTLS, you'd need this patch, yes.

hanno-arm commented:
The change looks good to me, and even appears to be mandatory looking at our own documentation of `mbedtls_cipher_update`, which states  `Writes as many block-sized blocks of data as possible to output.`. However, I think tests should be added exercising the new code-path.

AndrzejKurek commented:
I aggree with @hanno-arm on the test. @jethrogb - could you please add a test, checking if `mbedtls_cipher_update` returns more data when `ilen` equals `block_size` and `ctx->add_padding` is `NULL`?
Please see `tests/test_suite_cipher.padding.c` for test reference.
Could you also add a changelog entry under the `Changes` section, describing the change of behaviour, linking to this PR, and crediting you as the author of the contribution?
Thanks!

jethrogb commented:
I added additional parameters to the existing `enc_dec_buf_multipart` tests and added extra tests specifically for this case. When I run the new test suite against tip of `development`:
```
AES-128 CBC - Encrypt and decrypt 16 bytes in multiple parts with   FAILED
  (size_t)first_decrypt_output_len == outlen
  at line 449, suites/test_suite_cipher.function

AES-128 CBC - Encrypt and decrypt 16 bytes in multiple parts with   FAILED
  (size_t)second_decrypt_output_len == outlen
  at line 452, suites/test_suite_cipher.function

AES-128 CBC - Encrypt and decrypt 16 bytes in multiple parts with   FAILED
  (size_t)second_decrypt_output_len == outlen
  at line 452, suites/test_suite_cipher.function

AES-128 CBC - Encrypt and decrypt 16 bytes in multiple parts with   FAILED
  (size_t)second_decrypt_output_len == outlen
  at line 452, suites/test_suite_cipher.function

AES-128 CBC - Encrypt and decrypt 32 bytes in multiple parts with   FAILED
  (size_t)first_decrypt_output_len == outlen
  at line 449, suites/test_suite_cipher.function
```

mazimkhan commented:
retest

mazimkhan commented on library/cipher.c:
```diff
@@ -325,8 +325,10 @@ int mbedtls_cipher_update( mbedtls_cipher_context_t *ctx, const unsigned char *i
+        if( ( ctx->operation == MBEDTLS_DECRYPT && NULL != ctx->add_padding &&
                 ilen <= block_size - ctx->unprocessed_len ) ||
+            ( ctx->operation == MBEDTLS_DECRYPT && NULL == ctx->add_padding &&
+                ilen < block_size - ctx->unprocessed_len ) ||
```
For clarity following condition can be re-written:
```
 ( ctx->operation == MBEDTLS_DECRYPT && NULL != ctx->add_padding &&
     ilen <= block_size - ctx->unprocessed_len ) ||
 ( ctx->operation == MBEDTLS_DECRYPT && NULL == ctx->add_padding &&
     ilen < block_size - ctx->unprocessed_len )
```
as
```
( ctx->operation == MBEDTLS_DECRYPT &&  
    ( ( ilen < block_size - ctx->unprocessed_len ) || 
      ( ilen == block_size - ctx->unprocessed_len && NULL != ctx->add_padding ) ) )
```

Patater commented:
Mbed TLS isn't doing what it said it would, so this is a bug fix.

@AndrzejKurek Please update Changelog and make backports. Thanks.

gilles-peskine-arm commented on ChangeLog:
```diff
@@ -18,6 +18,8 @@ Bugfix
      Fixes #1040.
    * Fix C89 incompatibility in benchmark.c. Contributed by Brendan Shanks.
      #1353
+   * Return plaintext data sooner on unpadded decryption, as stated in
```
@AndrzejKurek Could you please make some small changes to the changelog entry, here and in the backports?

> Return plaintext data sooner on unpadded **CBC** decryption, as stated in **the** mbedtls_cipher_update documentation. Contributed by Andy Leiserson **in** #1180.

gilles-peskine-arm commented:
Note to gatekeepers: changelog needs tweaking (https://github.com/ARMmbed/mbedtls/pull/1180#discussion_r178285749), otherwise ok to merge.
